### PR TITLE
[recipe] fix: DAPO using KL in reward

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

Currently, the DAPO recipe will fail if we set `use_kl_in_reward` to True. This is because we need to calculate `response_mask`, `old_log_prob`, and/or `ref_log_prob` before we call `apply_kl_penalty` to rewards.

Although KL divergence is removed according to DAPO paper, it would still be neccessary to get this fixed, as `apply_kl_penalty` is already there in the code, also it can be good to support this to allow exploration in non-standard ways.

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: recipe/dapo/dapo_ray_trainer.py [hunk 1]
```diff
@@ -187,9 +215,14 @@ def fit(self):
                             new_batch = new_batch.union(reward_tensor)
 
                         # we combine with rule-based rm
-                        reward_tensor, reward_extra_infos_dict = compute_reward(new_batch, self.reward_fn)
+                        reward_tensor, format_tensor, correctness_tensor, length_tensor, reward_extra_infos_dict = (
+                            compute_reward(new_batch, self.global_steps, self.reward_fn)
+                        )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply recipe/dapo/dapo_ray_trainer.py]: fixed

File: recipe/dapo/dapo_ray_trainer.py [hunk 2]
```diff
@@ -51,6 +51,29 @@ class RayDAPOTrainer(RayPPOTrainer):
...
+            old_log_prob_metrics = {"actor/entropy": entropy_agg.detach().item()}
+            metrics.update(old_log_prob_metrics)
+            old_log_prob.batch.pop("entropys")
+            new_batch = new_batch.union(old_log_prob)
+
+        if self.use_reference_policy:
+            # compute reference log_prob
+            with marked_timer("ref", timing_raw, "olive"):
+                ref_log_prob = self.ref_policy_wg.compute_ref_log_prob(new_batch)
+                new_batch = new_batch.union(ref_log_prob)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply recipe/dapo/dapo_ray_trainer.py]: fixed

File: recipe/dapo/dapo_ray_trainer.py [hunk 3]
```diff
@@ -51,6 +51,32 @@ class RayDAPOTrainer(RayPPOTrainer):
...
+        if self.use_reference_policy:
+            # compute reference log_prob
+            with marked_timer("ref", timing_raw, "olive"):
+                if not self.ref_in_actor:
+                    ref_log_prob = self.ref_policy_wg.compute_ref_log_prob(new_batch)
+                else:
+                    ref_log_prob = self.actor_rollout_wg.compute_ref_log_prob(new_batch)
+                new_batch = new_batch.union(ref_log_prob)
+
+        return new_batch
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply recipe/dapo/dapo_ray_trainer.py]: fixed

File: recipe/dapo/dapo_ray_trainer.py [hunk 4]
```diff
@@ -51,6 +51,32 @@ class RayDAPOTrainer(RayPPOTrainer):
...
+        if self.use_reference_policy:
+            # compute reference log_prob
+            with marked_timer("ref", timing_raw, "olive"):
+                if not self.ref_in_actor:
+                    ref_log_prob = self.ref_policy_wg.compute_ref_log_prob(batch)
+                else:
+                    ref_log_prob = self.actor_rollout_wg.compute_ref_log_prob(batch)
+                batch = batch.union(ref_log_prob)
+
+        return batch
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply recipe/dapo/dapo_ray_trainer.py:78]: maybe it's better to do this fix in a separate PR.

File: recipe/entropy/entropy_ray_trainer.py [hunk 5]
```diff
@@ -48,6 +48,25 @@ class RayEntropyTrainer(RayPPOTrainer):
...
+        if self.use_reference_policy:
+            # compute reference log_prob
+            with simple_timer("ref", timing_raw):
+                if not self.ref_in_actor:
+                    ref_log_prob = self.ref_policy_wg.compute_ref_log_prob(batch)
+                else:
+                    ref_log_prob = self.actor_rollout_wg.compute_ref_log_prob(batch)
+                batch = batch.union(ref_log_prob)
+
+        return batch
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply recipe/entropy/entropy_ray_trainer.py:68]: maybe it's better to do this fix in a separate refactoring PR.

**Review by wuxibin89 — APPROVED** [review]
